content/multi-level-enums.md

```diff
@@ -0,0 +1,213 @@
+
+## Motivation
+
+Scala 3 introduced `enum` as a concise and type-safe way to define algebraic data types. However, it currently supports only flat enums. Many real-world use cases, such as domain modeling or UI state machines, naturally require **hierarchical enums**, where cases are grouped into logical families or categories.
+
+Consider modeling animals:
+
+```scala
+enum Animal:
+  case Dog, Cat, Sparrow, Penguin
+```
+
+This flat structure works, but grouping `Dog` and `Cat` under `Mammal`, and `Sparrow` and `Penguin` under `Bird` is more expressive and enables cleaner abstraction.
+
+
+## Proposed solution
+
+### High-level overview
+
+Allow `enum` cases to contain **nested `enum` definitions**, using a consistent indentation-based syntax.
+
+```scala
+enum Animal:
+  case enum Mammal:
+    case Dog, Cat
+  case enum Bird:
+    case Sparrow, Pinguin
+```
+
+Each nested `enum` case defines a group of related subcases. The **nested enum is itself a valid case** of the parent enum, and its members are **also valid cases**, allowing full exhaustivity and pattern matching.
+
+### Specification
+
+#### Enum Definition:
+
+```scala
+enum Animal:
+  case enum Mammal:
+    case Dog, Cat
+  case enum Bird:
+    case Sparrow, Pinguin
+  case Fish
+```
+
+- `case enum Mammal:` introduces a **sub-enum case**.
+- Nested cases (`Dog`, `Cat`) are **automatically part of the parent enum** (`Animal`), as well as part of the sub-enum (`Mammal`).
+
+#### Desugaring / Type Relationships
+
+The above syntax desugars to an enum tree with subtype relationships:
+
+```scala
+sealed trait Animal
+
+object Animal:
+  sealed trait Mammal extends Animal
+  object Mammal:
+    case object Dog extends Mammal 
+    case object Cat extends Mammal
+  
+  sealed trait Bird extends Animal
+  object Bird:
+   case object Sparrow extends Bird
+   case object Pinguin extends Bird
+  case object FIsh extends Animal
+```
+
+Results in:
+
+- `Mammal` and `Bird` are singleton enum cases of `Animal`
+- `Dog`, `Cat`, `Sparrow`, and `Pinguin` are **also** values of `Animal`, and they belong to `Mammal` and `Bird` respectively
+- Type relationships:
+  - `Dog <: Mammal <: Animal`
+  - `Cat <: Mammal <: Animal`
+  - `Fish <: Animal`
+  - etc.
+
+All leaf cases are usable as values of `Animal`, and the nested grouping allows matching at any level of the hierarchy.
+
+
+#### Pattern Matching
+
+Exhaustive pattern matching on `Animal` must cover all leaf cases:
+
+```scala
+def classify(a: Animal): String = a match
+  case Dog     => "a dog"
+  case Cat     => "a cat"
+  case Sparrow => "a bird"
+  case Pinguin => "a penguin"
+  case Fish    => "a fish"
+```
+
+Matching on intermediate enums is also allowed:
+
+```scala
+def isWarmBlooded(a: Animal): Boolean = a match
+  case Mammal => true     // Covers Dog, Cat
+  case Bird   => true     // Covers Sparrow, Pinguin
+  case Fish   => false
+```
+
+Matching on a **supercase** (e.g., `Mammal`) is shorthand for matching all its subcases.
```

Comment:
This does not make sense. Matching on `Mammal` itself already has meaning, which is to test if it is the `Mammal` companion `object` itself. That cannot be changed to means `case Cat | Dog`.